src/main/scala/com/fulcrumgenomics/sv/tools/AggregateSvPileup.scala

```diff
@@ -159,49 +158,34 @@ object AggregateSvPileup {
    * Converts a set of pairwise neighbor relationships to a set of clusters, where each cluster consists of breakpoints
    * that can be joined into a connected path via pairwise neighbor relationships.
    * @param pileupToNeighbors Map of pileup to its set of neighbors
-   * @param existingClusters An existing set of clusters to add to
    * @return The set of pileup clusters
    */
-  @tailrec
-  def toClusters(pileupToNeighbors: Map[BreakpointPileup, PileupGroup],
-                 existingClusters: Seq[PileupGroup] = Seq()): Seq[PileupGroup] = {
-    pileupToNeighbors.headOption match {
-      case None              => existingClusters
-      case Some((pileup, _)) =>
-        val (pileupGroup, remaining) = extractClusterFor(pileup, pileupToNeighbors)
-        toClusters(remaining, existingClusters :+ pileupGroup)
+  def toClusters(pileupToNeighbors: Map[BreakpointPileup, PileupGroup]): Seq[PileupGroup] = {
+    // The following finds connnected components in a breadth first manner
+    val visited    = scala.collection.mutable.HashSet[BreakpointPileup]()
+    val components = IndexedSeq.newBuilder[PileupGroup]
+    pileupToNeighbors.keys.foreach { rootPileup =>
+      if (!visited.contains(rootPileup)) {
+        val remaining = scala.collection.mutable.Queue[BreakpointPileup]()
+        val component = Set.newBuilder[BreakpointPileup]
+        remaining += rootPileup
+        while (remaining.nonEmpty) {
+          // get the next pileup to examine
+          val curPileup = remaining.dequeue()
+          if (!visited.contains(curPileup)) {
+            // add it to this component
+            component += curPileup
+            // add all unvisited neighbors
+            remaining ++= pileupToNeighbors(curPileup).filterNot(visited.contains)
```

Comment:
I wonder if we can avoid enqueueing the same pileup multiple times if we mark it as visited after we've enqueued it, then remove if on line 175.
```suggestion
            val unvisited = pileupToNeighbors(curPileup).filterNot(visited.contains)
            remaining ++= unvisited
            unvisited.foreach(visited.add)
```